src/Workspaces/Remote/Core/RemoteHostAssetReader.cs

```diff
@@ -0,0 +1,134 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Buffers;
+using System.IO.Pipelines;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Serialization;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.Remote;
+
+/// <summary>
+/// See <see cref="RemoteHostAssetWriter"/> for an explanation of the wire format we use when communicating assets
+/// between the host and our OOP server.  This implements the code for reading assets transmitted over the wire.  <see
+/// cref="RemoteHostAssetWriter"/> has the code for writing assets.
+/// </summary>
+internal readonly struct RemoteHostAssetReader<T, TArg>(
+    PipeReader pipeReader,
+    Checksum solutionChecksum,
+    int objectCount,
+    ISerializerService serializer,
+    Action<Checksum, T, TArg> callback,
+    TArg arg)
+{
+    private readonly PipeReader _pipeReader = pipeReader;
+    private readonly Checksum _solutionChecksum = solutionChecksum;
+    private readonly int _objectCount = objectCount;
+    private readonly ISerializerService _serializer = serializer;
+    private readonly Action<Checksum, T, TArg> _callback = callback;
+    private readonly TArg _arg = arg;
+
+    public ValueTask ReadDataAsync(CancellationToken cancellationToken)
+    {
+        // Suppress ExecutionContext flow for asynchronous operations operate on the pipe. In addition to avoiding
+        // ExecutionContext allocations, this clears the LogicalCallContext and avoids the need to clone data set by
+        // CallContext.LogicalSetData at each yielding await in the task tree.
+        //
+        // ⚠ DO NOT AWAIT INSIDE THE USING. The Dispose method that restores ExecutionContext flow must run on the
```

Comment:
>⚠ DO NOT AWAIT INSIDE THE USING. [](http://example.com/codeflow?start=11&length=32)

Every time I see this comment, I get a little frightened. For clarity, this is a limitation only on functionality directly in this method, not in methods invoked while inside the using, right?

Reply:
correct.

Reply:
Might be nice to update the comment to indicate this restriction is only for the current method, not for those it invokes in that context.

Reply:
done.